Snippet: Added Jest and Sinon for automated testing

mackiedrew opened the pull request:
# Commit Message
I added Jest because it is addressed in the issue #12 It is also the growing standard, and easy for individuals to pick up if they have never touched unit testing before. Sinon is a non-framework dependent JavaScript mocking library. This means either Jest or Sinon could be swapped out more easily later.

I have set the current coverage cut off to a typically low 30%. This is to make unit tests not absolutely required for participation in the project.

I've taken the liberty of writing the first function's tests to help test the test suite, and provide a template for people to checkout if they want to write laters. The function is pipe().

I've written robust documentation and put it in README.md.

# Summary of changes

- Prettier was run as a pre-commit hook, leading to a lot of changes.
- Added two dependecies: `jest`, `sinon`
- Added coverage to the .gitignore.
- Added documentation
- Added tests for pipe() function
- Added jest configuration in `package.json`
- Added jest setup in `jestSetup.js`

# Screenshots

You can see our only tests passing right now, I think I'll do some more tests next PR I made.
![tests w coverage](https://cloud.githubusercontent.com/assets/20718044/25626362/52feb1c6-2f2d-11e7-81ba-d7c08e9855df.png)

mackiedrew committed:
Added Jest and Sinon as for automated testing

I added Jest because it is addressed in the issue #12. It is also
the growing standard, and easy for individuals to pick up if they
have never touched unit testing before. Sinon is a non-framework
dependant JavaScript mocking library. This means either Jest or
Sinon could be swapped out more easilly later.

I have set the current coverage cut off to a typically low 30%.
This is to make unit tests not absolutely required for participation
in the project.

I've taken the libery of writing the first function's tests to
help test the test suite, and provide a template for people to
checkout if they want to write laters. The function is pipe().

I've written robust documentation and put it in README.md

teesloane commented:
This look excellent @mackiedrew. I really appreciate your hardwork on integrating testing notes into our readme. I will review and probably merge this tonight. Any chance you can post a beautiful screenshot of the tests you wrote going green?!

teesloane commented on README.md:
```diff
@@ -53,7 +53,63 @@ If you are a bit lost, the following terminology might help to clarify some aspe
+
+Currently only the **snippet** portion has automated testing setup.
+
+Now, the snippet is _not_ a JavaScript module, so the testing offered a challenge. The way it was solved is this case was to have a conditional export of each component that requires testing. You'll notice at the end of the snippet file `snippet/index.js` a series of lines that look like this:
```
If possible, write in present-tense, ex:

> Because the javascript snippet is not a module, testing requires conditional export of each component...

teesloane commented on README.md:
```diff
@@ -53,7 +53,63 @@ If you are a bit lost, the following terminology might help to clarify some aspe
+
+### Testing Stack
+
+1. [Jest](http://facebook.github.io/jest/): is our testing framework. It has some issues, but it's the new guy in town and relatively easy to pick.
```
Let's be kind to Jest, we don't need to say that it has some issues / is the new guy / easy to pick, a simple link is fine 😄

teesloane commented on README.md:
```diff
@@ -53,7 +53,63 @@ If you are a bit lost, the following terminology might help to clarify some aspe
+
+### How To Add Tests
+
+1. I've added a new function! Arg! What do I do?
```
This is good personal tone. I like this.

teesloane commented on README.md:
```diff
@@ -53,7 +53,63 @@ If you are a bit lost, the following terminology might help to clarify some aspe
+
+### Tips
+
+- Pure functions are much easier to test.
```
😻

teesloane commented on snippet/__tests__/pipe.test.js:
```diff
@@ -0,0 +1,17 @@
+    expect(typeof result).toBe("function");
+  });
+
+  it("all provided functions are called once when result function is called", () => {
```
try using present tense again :)

`it("invokes all functions passed in")...`...

teesloane commented on snippet/jestSetup.js:
```diff
@@ -0,0 +1,9 @@
+const sinon = require("sinon");
```
Can you move this file into our __tests__ folder?

also, can you rename `__tests__` > `test` ?

mackiedrew committed:
Requested changes are made after code review

Changes are as follows:

- README.md reflects new test folder name
- README.md is present tense where required
- Tense is corrected in test file.
- __tests__/ is now test/
- jestSetup.js is now in the snippet/test/ directory
- jestSetup.js is now called setup.js

mackiedrew commented:
Do you prefer pull request changes to be updated? Or do you prefer new commits for the changes?
